src/psyclone/psyir/transformations/arrayrange2loop_trans.py

```diff
@@ -0,0 +1,369 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2020, Science and Technology Facilities Council.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford, STFC Daresbury Lab
+
+'''Module providing a transformation from a PSyIR Array Range to a
+PSyIR Loop. This could be useful for e.g. performance reasons, to
+allow further transformations e.g. loop fusion or if the back-end does
+not support array ranges.
+
+'''
+
+from __future__ import absolute_import
+from psyclone.psyGen import Transformation
+from psyclone.psyir.transformations.transformation_error \
+    import TransformationError
+from psyclone.psyir.symbols import DataSymbol, INTEGER_TYPE
+from psyclone.psyir.nodes import Loop, Range, Reference, Array, Assignment, \
+    Node, Operation, BinaryOperation
+
+
+class ArrayRange2LoopTrans(Transformation):
+    '''Provides a transformation from a PSyIR Array Range to a PSyIR
+    Loop. For example:
+
+    >>> from psyclone.parse.algorithm import parse
+    >>> from psyclone.psyGen import PSyFactory
+    >>> api = "nemo"
+    >>> filename = "tra_adv.F90"
+    >>> ast, invoke_info = parse(filename, api=api)
```

Comment:
It would. However, at the moment the nemo front-end will create a NemoImplicitLoop, rather than an array with a Range. This means that it is not possible to use this transformation until this change is made unless we create an example directly in the PSyIR. A problem with creating an example directly in the PSyIR is that it is quite verbose (see the tests), so I don't think that is a good idea. As I will be removing NemoImplicitLoop soon, perhaps the example could be added then?